File: android/src/main/java/com/mux/reactnativeplayer/MuxVideoRecords.kt
package com.mux.reactnativeplayer

import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import com.mux.player.media.MediaItems
import com.mux.player.media.PlaybackResolution
import com.mux.player.media.RenditionOrder
import com.mux.stats.sdk.core.model.CustomData
import com.mux.stats.sdk.core.model.CustomerData
import com.mux.stats.sdk.core.model.CustomerPlayerData
import com.mux.stats.sdk.core.model.CustomerVideoData
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class MuxVideoClippingRecord : Record {
  @Field
  var assetStartTime: Double? = null

  @Field
  var assetEndTime: Double? = null
}

class MuxVideoMetadataRecord : Record {
  @Field
  var envKey: String? = null

  @Field
  var playerName: String? = null

  @Field
  var playerVersion: String? = null

  @Field
  var videoTitle: String? = null

  @Field
  var videoId: String? = null

  @Field
  var videoSeries: String? = null

  @Field
  var viewerUserId: String? = null

  @Field
  var customData: Map<String, String>? = null
}

class MuxVideoSourceRecord : Record {
  @Field
  var playbackId: String = ""

  @Field
  var playbackToken: String? = null

  @Field
  var drmToken: String? = null

  @Field
  var customDomain: String? = null

  @Field
  var minResolution: String? = null

  @Field
  var maxResolution: String? = null

  @Field
  var renditionOrder: String = "default"

  @Field
  var clipping: MuxVideoClippingRecord? = null

  @Field
  var metadata: MuxVideoMetadataRecord? = null

  val fingerprint: String
    get() = listOf(
      playbackId,
      playbackToken.orEmpty(),
      drmToken.orEmpty(),
      customDomain.orEmpty(),
      minResolution.orEmpty(),
      maxResolution.orEmpty(),
      renditionOrder,
      clipping?.assetStartTime?.toString().orEmpty(),
      clipping?.assetEndTime?.toString().orEmpty(),
      metadata?.envKey.orEmpty(),
      metadata?.playerName.orEmpty(),
      metadata?.playerVersion.orEmpty(),
      metadata?.videoTitle.orEmpty(),
      metadata?.videoId.orEmpty(),
      metadata?.videoSeries.orEmpty(),
      metadata?.viewerUserId.orEmpty(),
      metadata?.customData?.toSortedMap()?.entries?.joinToString { "${it.key}:${it.value}" }.orEmpty(),
    ).joinToString("|")

  fun toMediaItem(): MediaItem {
    val builder = MediaItems.builderFromMuxPlaybackId(
      playbackId = playbackId,
      maxResolution = maxResolution.toPlaybackResolution(),
      minResolution = minResolution.toPlaybackResolution(),
      renditionOrder = renditionOrder.toRenditionOrder(),
      assetStartTime = clipping?.assetStartTime,
      assetEndTime = clipping?.assetEndTime,
      domain = customDomain,
      playbackToken = playbackToken,
      drmToken = drmToken,
    )

    metadata?.videoTitle?.let { title ->
      builder.setMediaMetadata(
        MediaMetadata.Builder()
          .setTitle(title)
          .build()
      )
    }

    return builder.build()
  }

  fun toCustomerData(): CustomerData {
    val metadata = metadata
    return CustomerData().apply {
      customerPlayerData = CustomerPlayerData().apply {
        environmentKey = metadata?.envKey
        playerName = metadata?.playerName ?: "MuxReactNativePlayer"
        playerVersion = metadata?.playerVersion
        viewerUserId = metadata?.viewerUserId
      }
      customerVideoData = CustomerVideoData().apply {
        videoTitle = metadata?.videoTitle
        videoId = metadata?.videoId
        videoSeries = metadata?.videoSeries
      }
      customData = CustomData().apply {
        applyCustomData(metadata?.customData)
      }
    }
  }
}

private fun String?.toPlaybackResolution(): PlaybackResolution? {
  return when (this) {
    "480p" -> PlaybackResolution.LD_480
    "540p" -> PlaybackResolution.LD_540
    "720p" -> PlaybackResolution.HD_720
    "1080p" -> PlaybackResolution.FHD_1080
    "1440p" -> PlaybackResolution.QHD_1440
    "2160p" -> PlaybackResolution.FOUR_K_2160
    else -> null
  }
}

private fun String?.toRenditionOrder(): RenditionOrder? {
  return when (this) {
    "desc" -> RenditionOrder.Descending
    "default" -> RenditionOrder.Default
    else -> null
  }
}

private fun CustomData.applyCustomData(values: Map<String, String>?) {
  customData1 = values?.get("customData1")
  customData2 = values?.get("customData2")
  customData3 = values?.get("customData3")
  customData4 = values?.get("customData4")
  customData5 = values?.get("customData5")
  customData6 = values?.get("customData6")
  customData7 = values?.get("customData7")
  customData8 = values?.get("customData8")
  customData9 = values?.get("customData9")
  customData10 = values?.get("customData10")
}

File: ios/MuxVideoRecords.swift
import ExpoModulesCore
import Foundation
import MuxCore
import MuxPlayerSwift

struct MuxVideoClippingRecord: Record {
  @Field var assetStartTime: Double?
  @Field var assetEndTime: Double?
}

struct MuxCaptionStyleRecord: Record {
  @Field var textColor: String?
  @Field var backgroundColor: String?
  @Field var fontScale: Double?
}

struct MuxVideoMetadataRecord: Record {
  @Field var envKey: String?
  @Field var playerName: String?
  @Field var playerVersion: String?
  @Field var videoTitle: String?
  @Field var videoId: String?
  @Field var videoSeries: String?
  @Field var viewerUserId: String?
  @Field var customData: [String: String]?
}

struct MuxVideoSourceRecord: Record {
  @Field var playbackId: String = ""
  @Field var playbackToken: String?
  @Field var drmToken: String?
  @Field var thumbnailToken: String?
  @Field var storyboardToken: String?
  @Field var customDomain: String?
  @Field var minResolution: String?
  @Field var maxResolution: String?
  @Field var renditionOrder: String = "default"
  @Field var clipping: MuxVideoClippingRecord?
  @Field var metadata: MuxVideoMetadataRecord?

  var fingerprint: String {
    var values: [String] = []
    values.append(playbackId)
    values.append(playbackToken ?? "")
    values.append(drmToken ?? "")
    values.append(thumbnailToken ?? "")
    values.append(storyboardToken ?? "")
    values.append(customDomain ?? "")
    values.append(minResolution ?? "")
    values.append(maxResolution ?? "")
    values.append(renditionOrder)
    values.append(clipping?.assetStartTime.map { String($0) } ?? "")
    values.append(clipping?.assetEndTime.map { String($0) } ?? "")
    values.append(metadata?.envKey ?? "")
    values.append(metadata?.playerName ?? "")
    values.append(metadata?.playerVersion ?? "")
    values.append(metadata?.videoTitle ?? "")
    values.append(metadata?.videoId ?? "")
    values.append(metadata?.videoSeries ?? "")
    values.append(metadata?.viewerUserId ?? "")
    let customDataFingerprint = metadata?.customData?
      .sorted { $0.key < $1.key }
      .map { "\($0.key):\($0.value)" }
      .joined(separator: ",") ?? ""
    values.append(customDataFingerprint)

    return values.joined(separator: "|")
  }

  /// Build the Mux thumbnail URL used as lock-screen / Now Playing artwork.
  func artworkURL() -> URL? {
    guard !playbackId.isEmpty else {
      return nil
    }
    let host: String
    if let domain = blankToNil(customDomain) {
      host = "image.\(domain)"
    } else {
      host = "image.mux.com"
    }
    var components = URLComponents()
    components.scheme = "https"
    components.host = host
    components.path = "/\(playbackId)/thumbnail.jpg"
    if let token = blankToNil(thumbnailToken) {
      components.queryItems = [URLQueryItem(name: "token", value: token)]
    }
    return components.url
  }

  func toPlaybackOptions() -> PlaybackOptions {
    if let playbackToken, let drmToken {
      return PlaybackOptions(
        playbackToken: playbackToken,
        drmToken: drmToken,
        customDomain: blankToNil(customDomain)
      )
    }

    if let customDomain = blankToNil(customDomain), let playbackToken {
      return PlaybackOptions(customDomain: customDomain, playbackToken: playbackToken)
    }

    if let playbackToken {
      return PlaybackOptions(playbackToken: playbackToken)
    }

    if let customDomain = blankToNil(customDomain) {
      return PlaybackOptions(
        customDomain: customDomain,
        maximumResolutionTier: maxResolution.toMaxResolutionTier(),
        minimumResolutionTier: minResolution.toMinResolutionTier(),
        renditionOrder: renditionOrder.toRenditionOrder()
      )
    }

    if let clipping = clipping?.toInstantClipping() {
      return PlaybackOptions(
        maximumResolutionTier: maxResolution.toMaxResolutionTier(),
        minimumResolutionTier: minResolution.toMinResolutionTier(),
        renditionOrder: renditionOrder.toRenditionOrder(),
        clipping: clipping
      )
    }

    return PlaybackOptions(
      maximumResolutionTier: maxResolution.toMaxResolutionTier(),
      minimumResolutionTier: minResolution.toMinResolutionTier(),
      renditionOrder: renditionOrder.toRenditionOrder()
    )
  }

  func toMonitoringOptions() -> MonitoringOptions {
    let customerData = toCustomerData()
    let playerName = blankToNil(metadata?.playerName) ?? "MuxReactNativePlayer"
    return MonitoringOptions(
      customerData: customerData,
      playerName: playerName,
      automaticErrorTracking: true
    )
  }

  func toCustomerData() -> MUXSDKCustomerData {
    let customerData = MUXSDKCustomerData()

    let customerPlayerData = MUXSDKCustomerPlayerData()
    customerPlayerData.environmentKey = blankToNil(metadata?.envKey)
    customerPlayerData.playerName = blankToNil(metadata?.playerName) ?? "MuxReactNativePlayer"
    customerPlayerData.playerVersion = blankToNil(metadata?.playerVersion)
    customerPlayerData.viewerUserId = blankToNil(metadata?.viewerUserId)
    customerData.customerPlayerData = customerPlayerData

    let customerVideoData = MUXSDKCustomerVideoData()
    customerVideoData.videoTitle = blankToNil(metadata?.videoTitle)
    customerVideoData.videoId = blankToNil(metadata?.videoId)
    customerVideoData.videoSeries = blankToNil(metadata?.videoSeries)
    customerData.customerVideoData = customerVideoData

    let customData = MUXSDKCustomData()
    customData.customData1 = metadata?.customData?["customData1"]
    customData.customData2 = metadata?.customData?["customData2"]
    customData.customData3 = metadata?.customData?["customData3"]
    customData.customData4 = metadata?.customData?["customData4"]
    customData.customData5 = metadata?.customData?["customData5"]
    customData.customData6 = metadata?.customData?["customData6"]
    customData.customData7 = metadata?.customData?["customData7"]
    customData.customData8 = metadata?.customData?["customData8"]
    customData.customData9 = metadata?.customData?["customData9"]
    customData.customData10 = metadata?.customData?["customData10"]
    customerData.customData = customData

    return customerData
  }
}

private func blankToNil(_ value: String?) -> String? {
  guard let value, !value.isEmpty else {
    return nil
  }
  return value
}

private extension MuxVideoClippingRecord {
  func toInstantClipping() -> InstantClipping? {
    switch (assetStartTime, assetEndTime) {
    case let (start?, end?):
      return InstantClipping(assetStartTimeInSeconds: start, assetEndTimeInSeconds: end)
    case let (start?, nil):
      return InstantClipping(assetStartTimeInSeconds: start)
    case let (nil, end?):
      return InstantClipping(assetEndTimeInSeconds: end)
    default:
      return nil
    }
  }
}

private extension Optional where Wrapped == String {
  func toMaxResolutionTier() -> MaxResolutionTier {
    switch self {
    case "720p":
      return .upTo720p
    case "1080p":
      return .upTo1080p
    case "1440p":
      return .upTo1440p
    case "2160p":
      return .upTo2160p
    default:
      return .default
    }
  }

  func toMinResolutionTier() -> MinResolutionTier {
    switch self {
    case "480p":
      return .atLeast480p
    case "540p":
      return .atLeast540p
    case "720p":
      return .atLeast720p
    case "1080p":
      return .atLeast1080p
    case "1440p":
      return .atLeast1440p
    case "2160p":
      return .atLeast2160p
    default:
      return .default
    }
  }
}

private extension String {
  func toRenditionOrder() -> RenditionOrder {
    switch self {
    case "desc":
      return .descending
    default:
      return .default
    }
  }
}
